pushy/src/main/java/com/turo/pushy/apns/proxy/ProxyLocator.java

```diff
@@ -0,0 +1,48 @@
+package com.turo.pushy.apns.proxy;
+
+import java.net.InetSocketAddress;
+import java.net.Proxy;
+import java.net.ProxySelector;
+import java.net.SocketAddress;
+import java.net.URI;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Helper class to look for proxy configurations.
+ *
+ * @see <a href="https://docs.oracle.com/javase/7/docs/api/java/net/doc-files/net-properties.html#Proxies">The Java documentation on proxies.</a>
+ * @since 0.14
+ */
+class ProxyLocator {
+
+    private static final Logger log = LoggerFactory.getLogger(ProxyLocator.class);
+
+    /**
+     * Returns the {@link SocketAddress} for any system-wide {@link Proxy} of a given {@link Proxy.Type} for a given {@link URI}.
+     *
+     * @param uri the URI to find any system proxy for.
+     * @param proxyType the types of proxy to find for the URI.
+     * @return the {@link SocketAddress} of any <code>Proxy</code> found, or <code>null</code> if there were none.
+     * @since 0.14
+     */
+    static SocketAddress getProxyAddressForUri(final URI uri, final Proxy.Type proxyType) {
+        final ProxySelector defaultProxySelector = ProxySelector.getDefault();
+        final List<Proxy> proxiesForUri = defaultProxySelector.select(uri);
+        log.debug("Proxies for URI \"{}\" were {}", uri, proxiesForUri);
+
+        for (java.net.Proxy proxy : proxiesForUri) {
+            if (proxy.type() == proxyType) {
+                InetSocketAddress proxyAddress = (InetSocketAddress) proxy.address();
+                if (proxyAddress.isUnresolved()) {
```

Comment:
Sorry to be coming back to this so late, but can you say more about what's going on with this `isUnresolved` check?

Reply:
That could certainly do with a comment to explain, sorry!

It came from problems with the network we were deploying into, where the name was known but had not been resolved, which IIRC then caused trouble in actually using the proxy. Using https://docs.oracle.com/javase/8/docs/api/java/net/InetSocketAddress.html#InetSocketAddress-java.lang.String-int- is an attempt to force resolution.